File: tests/wasm/test_aroonosc.js
/**
 * WASM binding tests for Aroon Oscillator indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN 
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
    
    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32' 
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
        
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }
    
    testData = loadTestData();
});

test('AroonOsc partial params', () => {
    
    const highArray = new Float64Array(testData.high);
    const lowArray = new Float64Array(testData.low);
    
    
    const result = wasm.aroonosc_js(highArray, lowArray, 20);
    
    assert.strictEqual(result.length, highArray.length);
});

test('AroonOsc accuracy', async () => {
    
    const highArray = new Float64Array(testData.high);
    const lowArray = new Float64Array(testData.low);
    
    
    const expectedLastFive = [-50.0, -50.0, -50.0, -50.0, -42.8571];
    
    const result = wasm.aroonosc_js(highArray, lowArray, 14); 
    
    assert(result.length >= 5, "Not enough Aroon Osc values");
    assert.strictEqual(result.length, highArray.length);
    
    
    const last5 = result.slice(-5);
    
    assertArrayClose(
        last5,
        expectedLastFive,
        0.01, 
        "Aroon Osc last 5 values mismatch"
    );
});

test('AroonOsc default params', () => {
    
    const highArray = new Float64Array(testData.high);
    const lowArray = new Float64Array(testData.low);
    
    const result = wasm.aroonosc_js(highArray, lowArray, 14);
    assert.strictEqual(result.length, highArray.length);
});

test('AroonOsc zero length', () => {
    
    const high = new Float64Array([10.0, 20.0, 30.0]);
    const low = new Float64Array([5.0, 15.0, 25.0]);
    
    assert.throws(() => {
        wasm.aroonosc_js(high, low, 0);
    }, /Invalid length/);
});

test('AroonOsc length exceeds data', () => {
    
    const high = new Float64Array([10.0, 20.0, 30.0]);
    const low = new Float64Array([5.0, 15.0, 25.0]);
    
    assert.throws(() => {
        wasm.aroonosc_js(high, low, 10);
    }, /Not enough data/);
});

test('AroonOsc mismatched arrays', () => {
    
    const high = new Float64Array([10.0, 20.0, 30.0]);
    const low = new Float64Array([5.0, 15.0]); 
    
    assert.throws(() => {
        wasm.aroonosc_js(high, low, 14);
    }, /High and low arrays must have same length/);
});

test('AroonOsc NaN handling', () => {
    
    const highArray = new Float64Array(testData.high);
    const lowArray = new Float64Array(testData.low);
    
    const result = wasm.aroonosc_js(highArray, lowArray, 14);
    
    
    if (result.length > 50) {
        for (let i = 50; i < result.length; i++) {
            assert(!isNaN(result[i]), `Expected no NaN after index ${i}, but found NaN`);
        }
    }
});

test('AroonOsc reinput', () => {
    
    const highArray = new Float64Array(testData.high);
    const lowArray = new Float64Array(testData.low);
    
    
    const firstResult = wasm.aroonosc_js(highArray, lowArray, 10);
    const firstResultArray = new Float64Array(firstResult);
    
    
    const secondResult = wasm.aroonosc_js(firstResultArray, firstResultArray, 5);
    assert.strictEqual(secondResult.length, firstResult.length);
    
    
    for (let i = 20; i < secondResult.length; i++) {
        assert(!isNaN(secondResult[i]), `Found unexpected NaN at index ${i}`);
    }
});

test('AroonOsc batch single parameter set', () => {
    
    const highArray = new Float64Array(testData.high);
    const lowArray = new Float64Array(testData.low);
    
    
    const batchResult = wasm.aroonosc_batch_js(
        highArray,
        lowArray,
        14, 14, 0  
    );
    
    
    const singleResult = wasm.aroonosc_js(highArray, lowArray, 14);
    
    
    assert.strictEqual(batchResult.length, singleResult.length);
    
    assertArrayClose(batchResult, singleResult, 1e-10, "Batch vs single mismatch");
});

test('AroonOsc batch multiple lengths', () => {
    
    const high = testData.high.slice(0, 100);
    const low = testData.low.slice(0, 100);
    const highArray = new Float64Array(high);
    const lowArray = new Float64Array(low);
    
    
    const batchResult = wasm.aroonosc_batch_js(
        highArray,
        lowArray,
        10, 18, 4  
    );
    
    
    assert.strictEqual(batchResult.length, 3 * 100);
    
    
    const lengths = [10, 14, 18];
    
    for (let i = 0; i < lengths.length; i++) {
        const rowStart = i * 100;
        const rowEnd = rowStart + 100;
        const row = batchResult.slice(rowStart, rowEnd);
        
        const singleResult = wasm.aroonosc_js(highArray, lowArray, lengths[i]);
        
        assertArrayClose(
            row, 
            singleResult, 
            1e-10, 
            `Length ${lengths[i]} mismatch`
        );
    }
});

test('AroonOsc batch metadata', () => {
    
    const metadata = wasm.aroonosc_batch_metadata_js(
        10, 20, 2  
    );
    
    
    assert.strictEqual(metadata.length, 6);
    
    
    const expected = [10, 12, 14, 16, 18, 20];
    for (let i = 0; i < expected.length; i++) {
        assert.strictEqual(metadata[i], expected[i]);
    }
});

test('AroonOsc batch full parameter sweep', () => {
    
    const high = testData.high.slice(0, 50);
    const low = testData.low.slice(0, 50);
    const highArray = new Float64Array(high);
    const lowArray = new Float64Array(low);
    
    const batchResult = wasm.aroonosc_batch_js(
        highArray,
        lowArray,
        10, 20, 5  
    );
    
    const metadata = wasm.aroonosc_batch_metadata_js(10, 20, 5);
    
    
    assert.strictEqual(metadata.length, 3);
    assert.strictEqual(batchResult.length, 3 * 50); 
});

test('AroonOsc batch edge cases', () => {
    
    const data = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]);
    
    
    const singleBatch = wasm.aroonosc_batch_js(
        data,
        data,
        5, 5, 1
    );
    
    assert.strictEqual(singleBatch.length, 15); 
    
    
    assert.throws(() => {
        wasm.aroonosc_batch_js(
            new Float64Array([]),
            new Float64Array([]),
            10, 20, 2
        );
    });
});


test('AroonOsc batch - new ergonomic API with single parameter', () => {
    
    const highArray = new Float64Array(testData.high);
    const lowArray = new Float64Array(testData.low);
    
    const result = wasm.aroonosc_batch(highArray, lowArray, {
        length_range: [14, 14, 0]
    });
    
    
    assert(result.values, 'Should have values array');
    assert(result.combos, 'Should have combos array');
    assert(typeof result.rows === 'number', 'Should have rows count');
    assert(typeof result.cols === 'number', 'Should have cols count');
    
    
    assert.strictEqual(result.rows, 1);
    assert.strictEqual(result.cols, highArray.length);
    assert.strictEqual(result.combos.length, 1);
    assert.strictEqual(result.values.length, highArray.length);
    
    
    const combo = result.combos[0];
    assert.strictEqual(combo.length, 14);
    
    
    const oldResult = wasm.aroonosc_js(highArray, lowArray, 14);
    
    assertArrayClose(result.values, oldResult, 1e-10, "New vs old API mismatch");
});

test('AroonOsc batch - new API with multiple parameters', () => {
    
    const high = testData.high.slice(0, 50);
    const low = testData.low.slice(0, 50);
    const highArray = new Float64Array(high);
    const lowArray = new Float64Array(low);
    
    const result = wasm.aroonosc_batch(highArray, lowArray, {
        length_range: [10, 20, 5]  
    });
    
    
    assert.strictEqual(result.rows, 3);
    assert.strictEqual(result.cols, 50);
    assert.strictEqual(result.combos.length, 3);
    assert.strictEqual(result.values.length, 150); 
    
    
    const expectedLengths = [10, 15, 20];
    
    for (let i = 0; i < expectedLengths.length; i++) {
        assert.strictEqual(result.combos[i].length, expectedLengths[i]);
    }
    
    
    for (let i = 0; i < result.rows; i++) {
        const rowStart = i * result.cols;
        const rowEnd = rowStart + result.cols;
        const row = result.values.slice(rowStart, rowEnd);
        
        const oldResult = wasm.aroonosc_js(highArray, lowArray, result.combos[i].length);
        
        assertArrayClose(row, oldResult, 1e-10, `Row ${i} mismatch`);
    }
});

test('AroonOsc batch - new API error handling', () => {
    const highArray = new Float64Array(testData.high.slice(0, 10));
    const lowArray = new Float64Array(testData.low.slice(0, 10));
    
    
    assert.throws(() => {
        wasm.aroonosc_batch(highArray, lowArray, {
            length_range: [10, 20] 
        });
    }, /Invalid config/);
    
    
    assert.throws(() => {
        wasm.aroonosc_batch(highArray, lowArray, {
            
        });
    }, /Invalid config/);
    
    
    assert.throws(() => {
        wasm.aroonosc_batch(highArray, lowArray, {
            length_range: "invalid"
        });
    }, /Invalid config/);
});

test.after(() => {
    console.log('AroonOsc WASM tests completed');
});

File: tests/python/test_aroonosc.py
"""
Python binding tests for Aroon Oscillator indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import my_project


def load_test_data():
    """Load the standard test data"""
    data = np.loadtxt(
        'src/data/2018-09-01-2024-Bitfinex_Spot-4h.csv',
        delimiter=',',
        skiprows=1,
        usecols=(1, 3, 4, 2),  
        unpack=True
    )
    
    return tuple(np.ascontiguousarray(arr) for arr in data)


def test_aroonosc_partial_params():
    """Test Aroon Oscillator with partial parameters - mirrors check_aroonosc_partial_params"""
    _, high, low, _ = load_test_data()
    
    
    result = my_project.aroonosc(high, low, length=20)
    
    assert len(result) == len(high)
    assert isinstance(result, np.ndarray)
    
    
    for i in range(20):
        assert np.isnan(result[i])


def test_aroonosc_accuracy():
    """Test Aroon Oscillator matches expected values from Rust tests - mirrors check_aroonosc_accuracy"""
    _, high, low, _ = load_test_data()
    
    
    result = my_project.aroonosc(high, low)  
    
    
    expected_last_five = [-50.0, -50.0, -50.0, -50.0, -42.8571]
    
    assert len(result) >= 5, "Not enough Aroon Osc values"
    
    
    last_five = result[-5:]
    for i, (actual, expected) in enumerate(zip(last_five, expected_last_five)):
        assert abs(actual - expected) < 1e-2, \
            f"Aroon Osc mismatch at index {i}: expected {expected}, got {actual}"
    
    
    length = 14
    for i in range(length, len(result)):
        if not np.isnan(result[i]):
            assert np.isfinite(result[i]), f"Aroon Osc should be finite after enough data at index {i}"


def test_aroonosc_zero_length():
    """Test Aroon Oscillator fails with zero length - mirrors Rust test for zero period"""
    high = np.array([10.0, 20.0, 30.0])
    low = np.array([5.0, 15.0, 25.0])
    
    with pytest.raises(ValueError, match="Invalid length"):
        my_project.aroonosc(high, low, length=0)


def test_aroonosc_length_exceeds_data():
    """Test Aroon Oscillator fails when length exceeds data length - mirrors Rust test"""
    high = np.array([10.0, 20.0, 30.0])
    low = np.array([5.0, 15.0, 25.0])
    
    with pytest.raises(ValueError, match="Not enough data"):
        my_project.aroonosc(high, low, length=10)


def test_aroonosc_mismatched_arrays():
    """Test Aroon Oscillator fails with mismatched high/low arrays"""
    high = np.array([10.0, 20.0, 30.0])
    low = np.array([5.0, 15.0])  
    
    with pytest.raises(ValueError, match="High and low arrays must have same length"):
        my_project.aroonosc(high, low)


def test_aroonosc_nan_handling():
    """Test Aroon Oscillator handles NaN values correctly - mirrors check_aroonosc_nan_handling"""
    _, high, low, _ = load_test_data()
    
    result = my_project.aroonosc(high, low)
    
    
    if len(result) > 50:
        for i in range(50, len(result)):
            assert not np.isnan(result[i]), f"Expected no NaN after index {i}, but found NaN"


def test_aroonosc_streaming():
    """Test streaming Aroon Oscillator implementation"""
    _, high, low, _ = load_test_data()
    
    
    stream = my_project.AroonOscStream(length=14)
    
    
    stream_results = []
    for h, l in zip(high, low):
        val = stream.update(h, l)
        stream_results.append(val if val is not None else np.nan)
    
    
    batch_result = my_project.aroonosc(high, low, length=14)
    
    
    assert len(stream_results) == len(batch_result)
    
    
    
    
    
    
    warmup = 14
    assert all(np.isnan(v) for v in stream_results[:warmup]), \
        "Stream should return NaN during warmup"
    assert any(not np.isnan(v) for v in stream_results[warmup:]), \
        "Stream should produce values after warmup"
    
    
    
    
    
    
    
    


def test_aroonosc_batch_single_params():
    """Test batch with single parameter combination"""
    _, high, low, _ = load_test_data()
    
    
    batch_result = my_project.aroonosc_batch(
        high,
        low,
        length_range=(14, 14, 0)  
    )
    
    
    assert 'values' in batch_result
    assert 'lengths' in batch_result
    
    values = batch_result['values']
    lengths = batch_result['lengths']
    
    
    assert values.shape[0] == 1
    assert values.shape[1] == len(high)
    assert len(lengths) == 1
    assert lengths[0] == 14
    
    
    single_result = my_project.aroonosc(high, low, length=14)
    np.testing.assert_array_almost_equal(values[0], single_result, decimal=10)


def test_aroonosc_batch_multiple_lengths():
    """Test batch with multiple length values"""
    _, high, low, _ = load_test_data()
    
    
    batch_result = my_project.aroonosc_batch(
        high,
        low,
        length_range=(10, 18, 4)
    )
    
    values = batch_result['values']
    lengths = batch_result['lengths']
    
    
    assert values.shape[0] == 3
    assert values.shape[1] == len(high)
    assert len(lengths) == 3
    assert list(lengths) == [10, 14, 18]
    
    
    for i, length in enumerate(lengths):
        single_result = my_project.aroonosc(high, low, length=int(length))
        np.testing.assert_array_almost_equal(values[i], single_result, decimal=10,
                                           err_msg=f"Mismatch for length={length}")


def test_aroonosc_batch_kernel_options():
    """Test batch with different kernel options"""
    data = load_test_data()
    _, high, low, _ = data[0][:100], data[1][:100], data[2][:100], data[3][:100]  
    
    
    batch_scalar = my_project.aroonosc_batch(
        high,
        low,
        length_range=(10, 14, 2),
        kernel='scalar'
    )
    
    
    batch_auto = my_project.aroonosc_batch(
        high,
        low,
        length_range=(10, 14, 2),
        kernel=None  
    )
    
    
    np.testing.assert_array_almost_equal(
        batch_scalar['values'],
        batch_auto['values'],
        decimal=10
    )


def test_aroonosc_with_edge_cases():
    """Test Aroon Oscillator with edge case values"""
    
    high = np.full(50, 100.0)
    low = np.full(50, 100.0)
    
    result = my_project.aroonosc(high, low, length=14)
    
    
    for i in range(14, len(result)):
        assert abs(result[i]) < 1e-10, f"Expected 0 for constant values at index {i}, got {result[i]}"
    
    
    high = np.arange(50, dtype=float) + 100.0
    low = np.full(50, 100.0)
    
    result = my_project.aroonosc(high, low, length=14)
    
    
    for i in range(20, len(result)):
        assert result[i] > 0, f"Expected positive value for uptrend at index {i}, got {result[i]}"


def test_aroonosc_batch_empty_data():
    """Test batch fails with empty data"""
    high = np.array([])
    low = np.array([])
    
    with pytest.raises(ValueError):
        my_project.aroonosc_batch(high, low, length_range=(10, 20, 2))


def test_aroonosc_reinput():
    """Test Aroon Oscillator with reinput of its own output - mirrors Rust test"""
    _, high, low, _ = load_test_data()
    
    
    first_result = my_project.aroonosc(high, low, length=10)
    
    
    second_result = my_project.aroonosc(first_result, first_result, length=5)
    
    assert len(second_result) == len(first_result)
    
    
    for i in range(20, len(second_result)):
        assert not np.isnan(second_result[i])